Display error message on overlay when DAS is disconnected

nix-chu opened the pull request:
## Description

WOW ok I finished this PR quicker than I expected. GReat lol

When this PR is merged, it should:
- Display error message that the DAS is not connected to the overlay, or the broker has not initiated. 
  - Error message should disappear **only** when DAS is connected.
  - If DAS disconnects during the 'race', then the error message will reappear. 
  - Reconnect delay should also be changed to every 10 seconds. 
- Display data titles, even though data is not being sent to the overlay, when it didn't before.
- Add comment headers for the `Overlay` constructor in hopes that this will be more readable and organised.

The following changes are only applied to `overlay_new.py` at the moment. But I'm open to changing these for the other overlays if needed. 

ALSO, I'm open to any name changes for the new class `DASDisconnectMessage`, because it's quite long. But if everyone is cool with this, then I'm cool with it. 

## Screenshots

Initial version:
![image](https://user-images.githubusercontent.com/58575113/93505262-fb343980-f95d-11ea-840e-724ef2e8a728.png)

Updated version:
![image](https://user-images.githubusercontent.com/58575113/93604830-45262980-fa09-11ea-80f2-73d65b63591e.png)

## Steps to Test
Ensure that a `mosquitto` broker is **not** running in a terminal nor in the background. 

1. Run `poetry install`.
2. Run `poetry shell`.
3. Run `python overlay_new.py`.

To test if error message will disappear:

4. Run `mosquitto` in a terminal. Error message should disappear as soon as overlay is connected.

hallgchris commented on overlay.py:
```diff
@@ -50,6 +50,7 @@ def __init__(self, bike, width=1280, height=740, bg: str = None):
         self.client.on_connect = self._on_connect
         self.client.on_disconnect = self.on_disconnect
         self.client.on_log = self.on_log
+        self.client.is_connected = False  # Default is false until connected
```
It's best that we don't write new attributes to objects like this, especially when they're not one we've made and understand the internals of! If we were to go with keeping track of whether we're connected on our own, then we'd be best to stick with our own separate variable. However, lucky for us, the `paho-mqtt` client can tell us on it's own whether it's connected!

It's not listed in the docs for some reason, but in the source code [here](https://github.com/eclipse/paho.mqtt.python/blob/d7592a64b55469035ab66160919561c30ae5f645/src/paho/mqtt/client.py#L1349) we can see there's a public `is_connected()` method we can use to check for connection instead. In fact, this method would have been replaced by your boolean on this line! This is one reason why we shouldn't write new attributes to objects like this :)

nix-chu replied:
Didn't realise this existed, thanks for sharing!

hallgchris commented on overlay.py:
```diff
@@ -187,8 +188,22 @@ def on_connect(self, client, userdata, flags, rc):
+            Overlay implementations may override for one-off operations."""
+        pass
+
+    def display_base_layer(self):
```
To be consistent with other existing methods `draw_base_layer` might be slightly better here (and below for the protected method)

nix-chu replied:
@kdav108 I brought up this same issue when I first looked at the Raspicam, so hopefully I'm able to share this knowledge with you too 😄 . This explanation also includes `update_data_layer()` & `_update_data_layer()` and `on_connect()` & `_on_connect()`.

Basically, `overlay.py` is the parent class and has multiple subclasses that are used for the camera. `display_base_layer()` covers code logic and methods that are shared between **all** camera overlays, so this method is defined in `overlay.py`.  Whereas, `_display_base_layer()` is only used to cover functionality that is **specific** to that overlay. It should be noted that as soon as `display_base_layer()` finishes executing, then `_display_base_layer()` should be called straight after. 

For example, `update_data_layer()` in `overlay.py` is used to catch any errors that may occur while the data layer is being updated. This should be the same for all overlays. But for `_update_data_layer()`, this should vary between overlays. `_update_data_layer()` in `overlay_all_stats.py` should update the data layer at the *top strip*, while `_update_data_layer()` in `overlay_new.py` should update the data layer on the corners. (Note that `overlay_new.py` is our newest one, so obviously this overlay is more desirable to the eye and the code is soooooo much cleaner.) 

I know I started explaining `display_base_layer()` and then used `update_data_layer()` as an example, which is probably not a good thing. But the way these internal functions work is the same.

nix-chu replied:
It should also be noted that the non-private method should not be overridden by the sub-classes. `super()` could also work here, but iirc Chris brought up the issue that this would mean that a developer would need to know to use `super()` every time a new overlay is made.

hallgchris commented on overlay_new.py:
```diff
@@ -58,16 +64,21 @@ def data_field_coord(x, y):
     def _update_data_layer(self):
         self.data_canvas.clear()
 
+        if not self.client.is_connected:
+            self.das_disconnect_message.draw_data(self.data_canvas)
```
Lets move this logic to inside `draw_data` so we don't have to give this class special treatment. Pass a reference to `self.client` into the class through it's constructor, and then whenever we call `draw_data` on the class, we'll be able to check `client.is_connected()` to see if anything should be actually drawn from within the class.

This will mean we can put the `DasDisconnectMessage` in `self.components` with the rest of the `Components` and draw it in the for loop below, super clean!

hallgchris replied:
In the current system, components are either drawn only once, or every frame, so unfortunately this wouldn't be possible (again, in the current implementation).

nix-chu replied:
> If I'm not wrong, can't we just call `self.das_disconnect_message.draw_data(self.data_canvas)` in the `self.on_disconnect` method fo the parent? It makes more sense calling it there rather than have this `self.client.is_connected` attribute to track it.

I just tried this now, but the problem is that `self.on_disconnect()` only runs when the MQTT was initially running and then it disconnected. So if the broker was disconnected to begin with, then this disconnect message would not display. But you are right when you say it definitely makes sense.

I'm leaning more towards Chris' solution since this would mean that this class can be run inside the for loop of `update_data_layer()` and `draw_base_layer()`, and therefore more maintanable. 

I'll try changing up both approaches and hopefully we come to a solution that we are all satisfied with.

hallgchris commented on overlay.py:
```diff
@@ -94,15 +96,13 @@ def connect(self, ip="192.168.100.100", port=1883):
-                prev_data_update = (
-                    0  # time that we last updated the data layer
-                )
+                prev_data_update = 0  # time when data layer was updated
```
While we're formatting this line, can you chuck the comment on the line above?

hallgchris commented on overlay.py:
```diff
@@ -187,8 +188,22 @@ def on_connect(self, client, userdata, flags, rc):
+        with self.exception_handler:
+            self._display_base_layer()
+
+    def _display_base_layer(self):
```
This can be abstract! Also, we shouldn't need `pass`, as the docstring is sufficient method content. (there's a few other extra `pass`es in this file, remove them too if you feel like it haha)

hallgchris commented on overlay.py:
```diff
@@ -50,6 +50,7 @@ def __init__(self, bike, width=1280, height=740, bg: str = None):
         self.client.on_connect = self._on_connect
         self.client.on_disconnect = self.on_disconnect
         self.client.on_log = self.on_log
+        self.client.is_connected = False  # Default is false until connected
```
Can you set `self.client.reconnect_delay_set(max_delay=10)` or something like that? By default, the MQTT client can wait up to 2 minutes before retrying connection which is a bit long. I don't think there's anything to loose by trying more regularly, it's not like we have hundreds of devices on our network all retrying

hallgchris reviewed (changes requested):
Great stuff, super quick too! A few suggestions to polish this PR up a bit :)

nix-chu commented on components/message.py:
```diff
@@ -25,10 +25,16 @@ def draw_data(self, canvas: Canvas, data: Data):
         display_str = f"Message: {message}"
+        self.display_message(canvas, display_str)
+
+    def display_message(
+        self, canvas: Canvas, message: str, colour: Colour = Colour.black
+    ):
+        """ Display message on data overlay by wrapping each line starting
+            from the top."""
```
All good! Haha, I thought I could get away with leaving the DAShboardMessage as Message.

khanguslee replied:
what a work of art

nix-chu replied:
Want to point out that this is a good idea, since I was struggling with a circular import error yesterday

khanguslee reviewed (commented):
Reviewed. Nothing else to add on as @hallgchris is a genius.

kdav108 reviewed (changes requested):
Been a bit lazy so haven't actually tried running this, but I've got a few suggestions just from reading the code. I'm also not that familiar with how all these files play together in the larger system so I might have got some comments wrong lol.

hallgchris commented on components/das_disconnect_message.py:
```diff
@@ -0,0 +1,29 @@
+
+    text_colour = Colour.red
+
+    def draw_data(self, canvas: Canvas):
```
Honestly I'm not sure, though it wouldn't be too hard to experiment. My guess would be that although python would still treat it as a subclass, although as a programmer I don't think you really could anymore when the APIs differ

hallgchris commented on components/das_disconnect_message.py:
```diff
@@ -0,0 +1,29 @@
+        connected to broker.
+        """
+        disconnected_message = (
+            "Message: DAS is disconnected. Connect to a broker."
```
Actually that's a really good point. Thinking about it, I reckon we should have a message that is more actionable by non-electrical people, such as "Connecting to DAS, ensure that it is turned on..."

nix-chu commented:
Key changes:
- Converted Message class into an abstract class, and created DAShboardMessage class, which follows OOP principles.
- Moved checking whether MQTT is connected inside the DASDisconnectMessage class, by passing in the `self.client` MQTT broker and using the MQTT method `client.is_connected()`.
- Changed MQTT's reconnect delay to every 10 seconds.
- Added comment headers in `Overlay.__init__` in hopes that this constructor will be more organised and readable.
- Rewrote the disconnect message to be more user friendly. 

![image](https://user-images.githubusercontent.com/58575113/93604720-1ad46c00-fa09-11ea-9bce-cb663cd2be4e.png)

hallgchris commented on components/message.py:
```diff
@@ -1,15 +1,23 @@
 
-class Message(Component):
-    """ Displays any existing messages at the top of the screen.
+class Message(Component, ABC):
```
Because `Component` is abstract, `Message` will actually inherit it's abstractness if we don't implement all the required methods. So, we can actually leave `draw_data` out entirely from this file, and not explicitly inherit `ABC`. Because subclasses of `Message` will also be subclasses of `Component`, and `Message` does not provide an implementation of `draw_data`, these subclasses will then be forced to implement `draw_data`.

hallgchris commented on components/dashboard_message.py:
```diff
@@ -0,0 +1,17 @@
+class DAShboardMessage(Message):
+    """Displays messages sent from the DAShboard onto the camera overlay.
+
+    Messages are only sent when the overlay is connected to the DAS.
```
"Messages can only be received when the overlay is connected to the DAS" might be more accurate.

hallgchris commented on components/__init__.py:
```diff
@@ -1,14 +1,14 @@
     "SpeedField",
     "CentrePower",
-    "Message",
+    "DAShboardMessage",
```
Include `Message` here as well! Will help below

nix-chu replied:
Is this file required when importing classes from a folder?

hallgchris replied:
Not required, but it can definitely make things neater! This defines what the module `components` will contain, meaning we can do things like `from components import DAShboardMessage` rather than needing to specify the exact file too with `from components.dashboard_message import DAShboardMessage`. It also defines what will be imported with `from components import *`, although this is generally not something we do.

hallgchris commented on components/das_disconnect_message.py:
```diff
@@ -1,29 +1,32 @@
 from components.message import Message
```
If you make the change to `__init__.py` above, you'll be able to simplify this to `from components import Message`

hallgchris commented on components/das_disconnect_message.py:
```diff
@@ -1,29 +1,32 @@
+        client: MQTT Client which is used to check if client is connected
     """
 
+    # These class variables must be consistent for every instance
```
To be pedantic, I don't no if this comment is necessary - this is exactly what a class variable is

nix-chu replied:
Ah I wrote that to explain why `text_colour` and `message` is a class variable rather than being instantiated as an instance variable.

nix-chu replied:
Hm i suppose this is obvious anyways, all good.

hallgchris commented on overlay_new.py:
```diff
@@ -62,23 +62,20 @@ def data_field_coord(x, y):
-        self.das_disconnect_message = DASDisconnectMessage()
 
     def on_connect(self, client, userdata, flags, rc):
         print("Connected with rc: {}".format(rc))
```
I've realised that we print this message for all overlays, and then do nothing else important on connection. Can you move this into the default implementation of `on_connect`, and then make the function non-abstract? Seems like we don't really have a need for it anymore with all this code cleaning up :)

hallgchris reviewed (changes requested):
Really close, just some minor comments now!

Also, I realised that the old overlays need updating to use `_update_data_layer`, and won't start currently :sweat_smile: Can you quickly update them?

hallgchris commented on components/__init__.py:
```diff
@@ -9,6 +9,7 @@
     "DataField",
     "SpeedField",
     "CentrePower",
+    "Message",
```
You're gonna have to import `Message` too. (`OverlayNew` won't even start right now lol)

hallgchris commented on components/message.py:
```diff
@@ -20,15 +28,18 @@ class Message(Component):
-            return
-        message = data.get_message()
-        display_str = f"Message: {message}"
+        pass
```
Following from https://github.com/monash-human-power/raspicam/pull/48#discussion_r491236379, we don't actually need this method specified here at all, the abstract method will be inherited from `Component`.

hallgchris reviewed (changes requested):
We're super close! Still haven't quite finished updating the old overlays, I believe all you need to do is rename `on_connect` to `_draw_base_layer`. You should be able to run them without errors :)

kdav108 commented on components/das_disconnect_message.py:
```diff
@@ -0,0 +1,33 @@
+class DASDisconnectMessage(Message):
+    """Class displays error message on overlay when DAS is not connected.
+
+    Message can only be received when the DAS is not connected. Will
```
`received` -> `displayed`?

kdav108 replied:
Actually on second thought we don't need this sentence since it conveys the same message as the one previous to it.

nix-chu replied:
Hm, I was thinking about this, but I wanted this to be clearly pointed out in the docstring, so that there's no misunderstanding in the future that the disconnected message would appear at a random different time **other** than when the DAS is disconnected. 

Maybe I could change the first line of the docstring to "Class will display an error message on the overlay that the DAS is not connected."

kdav108 reviewed (commented):
Just a few more comments (not very focussed on the technical aspects). Apart from that this looks great to me!

nix-chu commented on overlay.py:
```diff
@@ -169,7 +175,13 @@ def _on_connect(self, client, userdata, flags, rc):
-        self.backend.on_base_canvas_updated(self.base_canvas)
+        print("Connected with rc: {}".format(rc))
+
+    def on_connect(self, client, userdata, flags, rc):
```
Based on what Chris said https://github.com/monash-human-power/raspicam/pull/48#discussion_r491239207, this shouldn't be abstract for the moment. Since `overlay_new.py` does not override `on_connect()` because it doesn't need to anymore. However, this is currently used in the other overlays. The other overlays currently use `on_connect()` to draw the base layer. 

I imagine in the future, we will need to update the other overlays to start using `draw_base_layer()`. However, I don't think this is appropriate for this PR, specifically because the other overlays are very outdated compared to `overlay_new`. This should be a PR on its own, since the other overlays don't seem to be using `components`, and they definitely need a lot of refactoring. Combining those tasks would make this PR too unnecessarily big, in my opinion.

nix-chu replied:
Will definitely bring this up tomorrow with Chris and suggest making a PR to update the other overlays.

nix-chu replied:
Based on what Chris said in https://github.com/monash-human-power/raspicam/pull/48#discussion_r491239207, this shouldn't be abstract for the moment, because this would force the subclass overlays to implement this when they no longer need it. At the moment, the only overlay that uses `on_connect()` is `overlay_error` which is used for testing whether an error message will be sent to the DAShboard. 

I definitely think there needs to be a lot of refactoring done with the Raspicam, because of all these new changes. I will talk to Chris tomorrow about cleaning up a lot of the code in the other overlays, because I really love how `overlay_new` currently handles displaying components onto the overlay.

nix-chu replied:
Upon saying that, I won't do this in this PR now, because that would be too big and would need to be a PR on its own. Hopefully you understand.

kdav108 replied:
Yep my bad I didn't seem Chris's comment regrading this.

nix-chu replied:
Haha all good, I resolved it so there was no way you could've seen that. Not your fault 😄

hallgchris commented on overlay.py:
```diff
@@ -50,6 +51,7 @@ def __init__(self, bike, width=1280, height=740, bg: str = None):
         self.client.on_connect = self._on_connect
         self.client.on_disconnect = self.on_disconnect
```
I reckon it's still useful to have. Not all overlays display whether the display is connected to the DAS, so it is still useful for debugging those overlays. I don't think it causes any harm by staying :)

hallgchris reviewed (approved):
Looks good to me! Sorry about the delay, will try to keep things quicker during the break :)

kdav108 reviewed (approved):
Great job on this PR!
